### ISSUE TITLE
[PATCH] Locale::Maketext - speed and efficiency tweaks

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10508]
Migrated from [rt.perl.org#76674](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=76674) (status was 'resolved')

Searchable as RT76674$

**Comment by p5pRT** [comment]
### From @toddr

#### Created by @toddr



This patch resolves CPAN RT \#48118 of the same name\.

The change checks for \[ in the string before it attempts to compile it\. If \[ is not found\, it is able to bypass compile\. We've found that this change benchmarked 9% faster in our code\. 

<details><summary>Perl Info</summary>

```

Flags:
   category=library
   severity=low
   module=Locale::Maketext
   Type=Patch
   PatchStatus=HasPatch

Site configuration information for perl 5.12.1:

Configured by cPanel at Mon Jun  7 11:16:28 CDT 2010.

Summary of my perl5 (revision 5 version 12 subversion 1) configuration:

 Platform:
   osname=linux, osvers=2.6.18-194.3.1.el5, archname=x86_64-linux
   uname='linux rpmb-centos-50-64bit 2.6.18-194.3.1.el5 #1 smp thu may 13 13:08:30 edt 2010 x86_64 x86_64 x86_64 gnulinux '
   config_args='-des -Darchname=x86_64-linux -Dcc=/usr/local/cpanel/bin/gcc -Dcpp=/usr/local/cpanel/bin/gcc -E -DDEBUGGING=none -Doptimize=-Os -Dusemymalloc=y -Duseshrplib=true -Duselargefiles=yes -Duseposix=true -Dhint=recommended -Duseperlio=yes -Dccflags=-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Wl,-rpath -Wl,/usr/local/cpanel/lib64  -Dcppflags=-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Dldflags=-Wl,-rpath -Wl,/usr/local/cpanel/lib64 -L/usr/local/cpanel/lib64 -Dprefix=/usr/local/cpanel -Dsiteprefix=/usr/local/cpanel -Dsitebin=/usr/local/cpanel/bin -Dsitelib=/usr/local/cpanel/lib64/perl5/site_lib -Dprivlib=/usr/local/cpanel/lib64/perl5/5.12.1 -Dotherlibdirs=/var/cpanel/perl5/lib:/usr/local/cpanel/lib64/perl5/cpanel -Dman1dir=/usr/local/cpanel/share/man/man1 -Dman3dir=/usr/local/cpanel/share/man/man3 -Dsiteman1dir=/usr/local/cpanel/share/man/man1 -Dsiteman3dir=/usr/local/cpanel/share/man/man3 -Dcf_by=cPanel -Dmyhostname=localhost -Dperladmin=root@localhost -Dcf_email=support@cpanel.net -Di_dbm=/usr/local/cpanel/include -Di_gdbm=/usr/local/cpanel/include -Di_ndbm=/usr/local/cpanel/include -Ud_dosuid -Uuserelocatableinc -Umad -Uusethreads -Uusemultiplicity -Uusesocks -Uuselongdouble -Ui_db -Aldflags=-L/usr/local/cpanel/lib64 -L/usr/lib64 -L/lib64 -lgdbm -Dlocincpth=/usr/local/cpanel/include /usr/local/include  -Acflags=-fPIC -DPIC -m64 -I/usr/local/cpanel/include -Dlibpth=/usr/local/cpanel/lib64 /usr/local/lib64 /usr/local/lib /lib64 /usr/lib64  -Duse64bitint=yes -Duse64bitall=yes'
   hint=recommended, useposix=true, d_sigaction=define
   useithreads=undef, usemultiplicity=undef
   useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
   use64bitint=define, use64bitall=define, uselongdouble=undef
   usemymalloc=y, bincompat5005=undef
 Compiler:
   cc='/usr/local/cpanel/bin/gcc', ccflags ='-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Wl,-rpath -Wl,/usr/local/cpanel/lib64 -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/cpanel/include -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
   optimize='-Os',
   cppflags='-I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -I/usr/local/cpanel/include -L/usr/local/cpanel/lib64 -Wl,-rpath -Wl,/usr/local/cpanel/lib64 -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/cpanel/include -I/usr/local/include'
   ccversion='', gccversion='4.5.0', gccosandvers=''
   intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
   d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
   ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
   alignbytes=8, prototype=define
 Linker and Libraries:
   ld='/usr/local/cpanel/bin/gcc', ldflags ='-Wl,-rpath -Wl,/usr/local/cpanel/lib64 -L/usr/local/cpanel/lib64 -L/usr/local/cpanel/lib64 -L/usr/lib64 -L/lib64 -lgdbm -fstack-protector -L/usr/local/lib'
   libpth=/usr/local/cpanel/lib64 /usr/local/lib64 /usr/local/lib /lib64 /usr/lib64
   libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
   perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
   libc=/lib/libc-2.5.so, so=so, useshrplib=true, libperl=libperl.so
   gnulibc_version='2.5'
 Dynamic Linking:
   dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/usr/local/cpanel/lib64/perl5/5.12.1/x86_64-linux/CORE'
   cccdlflags='-fPIC', lddlflags='-shared -Os -L/usr/local/cpanel/lib64 -L/usr/lib64 -L/lib64 -L/usr/local/lib -fstack-protector'

Locally applied patches:
   cPanel Patches


@INC for perl 5.12.1:
   /usr/local/cpanel/lib64/perl5/site_lib/x86_64-linux
   /usr/local/cpanel/lib64/perl5/site_lib
   /usr/local/cpanel/lib64/perl5/5.12.1/x86_64-linux
   /usr/local/cpanel/lib64/perl5/5.12.1
   /var/cpanel/perl5/lib
   /usr/local/cpanel/lib64/perl5/cpanel
   .


Environment for perl 5.12.1:
   HOME=/home/toddr
   LANG=en_US.UTF-8
   LANGUAGE (unset)
   LD_LIBRARY_PATH (unset)
   LOGDIR (unset)
   PATH=/usr/local/cpanel/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/X11/bin:
   PERL_BADLANG (unset)
   SHELL=/bin/zsh


```
</details>

**Comment by p5pRT** [comment]
### From @toddr

<details><summary>Inline Patch</summary>

```diff
diff --git a/dist/Locale-Maketext/lib/Locale/Maketext.pm b/dist/Locale-Maketext/lib/Locale/Maketext.pm
index 929a70e..f23153a 100644
--- a/dist/Locale-Maketext/lib/Locale/Maketext.pm
+++ b/dist/Locale-Maketext/lib/Locale/Maketext.pm
@@ -146,8 +146,8 @@ sub failure_handler_auto {
     $handle->{'failure_lex'} ||= {};
     my $lex = $handle->{'failure_lex'};
 
-    my $value;
-    $lex->{$phrase} ||= ($value = $handle->_compile($phrase));
+     # tr/X// is a very fast "does this string have this character" check, so we don't even call _compile if we know it will end up being a SCALAR ref
+     my $value = $lex->{$phrase} ||= ($phrase !~ tr/[// ? \"$phrase" : $handle->_compile($phrase));
 
     # Dumbly copied from sub maketext:
     return ${$value} if ref($value) eq 'SCALAR';
@@ -214,11 +214,13 @@ sub maketext {
                 DEBUG and warn "  Found \"$phrase\" in $h_r\n";
                 unless(ref($value = $h_r->{$phrase})) {
                     # Nonref means it's not yet compiled.  Compile and replace.
-                    if ($handle->{'use_external_lex_cache'}) {
-                        $value = $handle->{'_external_lex_cache'}{$phrase} = $handle->_compile($value);
+                    if ($handle->{'use_external_lex_cache'}) {   
+                        # tr/X// is a very fast "does this string have this character" check, so we don't even call _compile if we know it will end up being a SCALAR ref
+                        $handle->{'_external_lex_cache'}{$phrase} = $value = ($value !~ tr/[// ? \"$value" : $handle->_compile($value));
                     }
                     else {
-                        $value = $h_r->{$phrase} = $handle->_compile($value);
+                        # tr/X// is a very fast "does this string have this character" check, so we don't even call _compile if we know it will end up being a SCALAR ref
+                        $h_r->{$phrase} = $value = ($value !~ tr/[// ? \"$value" : $handle->_compile($value));
                     }
                 }
                 last;
@@ -228,11 +230,13 @@ sub maketext {
             elsif($phrase !~ m/^_/s and ($handle->{'use_external_lex_cache'} ? ( exists $handle->{'_external_lex_cache'}{'_AUTO'} ? $handle->{'_external_lex_cache'}{'_AUTO'} : $h_r->{'_AUTO'} ) : $h_r->{'_AUTO'})) {
                 # it's an auto lex, and this is an autoable key!
                 DEBUG and warn "  Automaking \"$phrase\" into $h_r\n";
-                if ($handle->{'use_external_lex_cache'}) {
-                    $value = $handle->{'_external_lex_cache'}{$phrase} = $handle->_compile($phrase);
+                if ($handle->{'use_external_lex_cache'}) {   
+                    # tr/X// is a very fast "does this string have this character" check, so we don't even call _compile if we know it will end up being a SCALAR ref
+                    $handle->{'_external_lex_cache'}{$phrase} = $value = ($phrase !~ tr/[// ? \"$phrase" : $handle->_compile($phrase));
                 }
                 else {
-                    $value = $h_r->{$phrase} = $handle->_compile($phrase);
+                    # tr/X// is a very fast "does this string have this character" check, so we don't even call _compile if we know it will end up being a SCALAR ref
+                    $h_r->{$phrase} = $value = ($phrase !~ tr/[// ? \"$phrase" : $handle->_compile($phrase));
                 }
                 last;
             }
@@ -487,16 +491,14 @@ sub _compile {
     # This big scary routine compiles an entry.
     # It returns either a coderef if there's brackety bits in this, or
     #  otherwise a ref to a scalar.
-
-    my $target = ref($_[0]) || $_[0];
-
-    my(@code);
-    my(@c) = (''); # "chunks" -- scratch.
-    my $call_count = 0;
-    my $big_pile = '';
+    
+    # tr/X// is a very fast "does this string have this character" check, so we don't need to continue if we know it will end up being a SCALAR ref
+    return \"$_[1]" if $_[1] !~ tr/[//;
+    
+    my ($target,$call_count,$big_pile,@c,@code) = ( (ref($_[0]) || $_[0]), 0, '', '' );
     {
-        my $in_group = 0; # start out outside a group
-        my($m, @params); # scratch
+                                          # start out outside a group
+        my($in_group, $m, @params) = (0); # scratch
 
         while($_[1] =~  # Iterate over chunks.
             m/\G(
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Jul 22 11&#8203;:09&#8203;:33 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
> This patch resolves CPAN RT \#48118 of the same name\.
> 
> The change checks for \[ in the string before it attempts to compile
>    it\. If \[ is not found\, it is able to bypass compile\. We've found
>    that this change benchmarked 9% faster in our code\.

In my experience\, $a \!~ /\\\[/ is faster than $a \!~ y/\[//\, because the
former can stop when it finds a \[ and doesn’t have to keep count the way
y/// does\.

Using Steffen Mueller’s dumbbench script\, I confirmed that /\\\[/ takes
longer to compile\, but y/\[// takes longer to run\.

Also\, from reading the source \(not actually testing this\) it seems that
your patch stops an unmatched clasing bracket from being an error\. Am I
right about this?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @toddr


On Sep 26\, 2010\, at 7&#8203;:28 PM\, Father Chrysostomos via RT wrote&#8203;:

> On Thu Jul 22 11&#8203;:09&#8203;:33 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
>> This patch resolves CPAN RT \#48118 of the same name\.
>> 
>> The change checks for \[ in the string before it attempts to compile
>>   it\. If \[ is not found\, it is able to bypass compile\. We've found
>>   that this change benchmarked 9% faster in our code\.
> 
> In my experience\, $a \!~ /\\\[/ is faster than $a \!~ y/\[//\, because the
> former can stop when it finds a \[ and doesn’t have to keep count the way
> y/// does\.
> 
> Using Steffen Mueller’s dumbbench script\, I confirmed that /\\\[/ takes
> longer to compile\, but y/\[// takes longer to run\.
> 
> Also\, from reading the source \(not actually testing this\) it seems that
> your patch stops an unmatched clasing bracket from being an error\. Am I
> right about this?
> 

Thanks for your careful review of this patch\. You were right that the patch short circuits error conditions\. I've simplified the patch\. I'm pasting the commit message below since it's got all of my thoughts in it&#8203;:

&nbsp;    RT 76668 \- Check string to compile for chars ~\]\[ and return \\"$string" if not found\.
&nbsp;    This is a 250% speed improvement on strings which don't require compile and only a
&nbsp;    ~2% hit if they did need compiling\.
&nbsp;    
&nbsp;    Remove \\G since everything is being captured it has no value\. This means we don't
&nbsp;    have to worry about seting pos $string\_to\_compile = 0 to prevent the previous
&nbsp;    regex from affecting this one\. There is a negligible speed improvement removing
&nbsp;    the \\G

All tests continue to pass\.

FYI if you want to benchmark on your own\, this is the method I used to show the speedup&#8203;:

Testing compiled string performance&#8203;:
perl \-Ilib \-MBenchmark \-MLocale&#8203;::Maketext \-le 'package X;our @&#8203;ISA=qw\(Locale&#8203;::Maketext\);package X&#8203;::en;our @&#8203;ISA=qw\(X\);our %Lexicon = \(\_AUTO => 1\);package main;my $h=X\->get\_handle\(\); my $c = 1; timethis\(100000\, sub\{$h\->maketext\("bar $c \[\_1\]"\, $c\); $c\+\+\}\)'

Testing strings which don't need compile
perl \-Ilib \-MBenchmark \-MLocale&#8203;::Maketext \-le 'package X;our @&#8203;ISA=qw\(Locale&#8203;::Maketext\);package X&#8203;::en;our @&#8203;ISA=qw\(X\);our %Lexicon = \(\_AUTO => 1\);package main;my $h=X\->get\_handle\(\); my $c = 1; timethis\(800000\, sub\{$h\->maketext\("bar $c"\); $c\+\+\}\)'

**Comment by p5pRT** [comment]
### From @toddr

commit f97f0042fe33db2388213c4926fd8f2bcacf3ad7
Author&#8203;: Todd Rinaldo \<toddr@&#8203;cpan\.org>
Date&#8203;:   Tue Sep 28 11&#8203;:31&#8203;:20 2010 \-0500

&nbsp;    RT 76668 \- Check string to compile for chars ~\]\[ and return \\"$string" if not found\.
&nbsp;    This is a 250% speed improvement on strings which don't require compile and only a
&nbsp;    ~2% hit if they did need compiling\.
&nbsp;    
&nbsp;    Remove \\G since everything is being captured it has no value\. This means we don't
&nbsp;    have to worry about seting pos $string\_to\_compile = 0 to prevent the previous
&nbsp;    regex from affecting this one\. There is a negligible speed improvement removing
&nbsp;    the \\G

<details><summary>Inline Patch</summary>

```diff
diff --git a/dist/Locale-Maketext/lib/Locale/Maketext.pm b/dist/Locale-Maketext/lib/Locale/Maketext.pm
index cbfcb14..a685d6b 100644
--- a/dist/Locale-Maketext/lib/Locale/Maketext.pm
+++ b/dist/Locale-Maketext/lib/Locale/Maketext.pm
@@ -489,6 +489,13 @@ sub _compile {
     # It returns either a coderef if there's brackety bits in this, or
     #  otherwise a ref to a scalar.
 
+    my $string_to_compile = $_[1]; # There are taint issues using regex on @_ - perlbug 60378,27344
+
+    # The while() regex is more expensive than this check on strings that don't need a compile.
+    # this op causes a ~2% speed hit for strings that need compile and a 250% speed improvement
+    # on strings that don't need compiling.
+    return \"$string_to_compile" if($string_to_compile !~ m/[\[~\]]/ms); # return a string ref if chars [~] are not in the string
+
     my $target = ref($_[0]) || $_[0];
 
     my(@code);
@@ -499,10 +506,9 @@ sub _compile {
         my $in_group = 0; # start out outside a group
         my($m, @params); # scratch
 
-	my $string_to_compile = $_[1]; # There are taint issues using regex on @_ - perlbug 60378,27344
         while($string_to_compile =~  # Iterate over chunks.
-            m/\G(
-                [^\~\[\]]+  # non-~[] stuff
+            m/(
+                [^\~\[\]]+  # non-~[] stuff (Capture everything else here)
                 |
                 ~.       # ~[, ~], ~~, ~other
                 |
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Sep 28 10&#8203;:07&#8203;:11 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
> 
> On Sep 26\, 2010\, at 7&#8203;:28 PM\, Father Chrysostomos via RT wrote&#8203;:
> 
> > On Thu Jul 22 11&#8203;:09&#8203;:33 2010\, toddr@&#8203;cpanel\.net wrote&#8203;:
> >> This patch resolves CPAN RT \#48118 of the same name\.
> >>
> >> The change checks for \[ in the string before it attempts to compile
> >>   it\. If \[ is not found\, it is able to bypass compile\. We've found
> >>   that this change benchmarked 9% faster in our code\.
> >
> > In my experience\, $a \!~ /\\\[/ is faster than $a \!~ y/\[//\, because the
> > former can stop when it finds a \[ and doesn’t have to keep count the
>    way
> > y/// does\.
> >
> > Using Steffen Mueller’s dumbbench script\, I confirmed that /\\\[/
>    takes
> > longer to compile\, but y/\[// takes longer to run\.
> >
> > Also\, from reading the source \(not actually testing this\) it seems
>    that
> > your patch stops an unmatched clasing bracket from being an error\.
>    Am I
> > right about this?
> >
> 
> Thanks for your careful review of this patch\. You were right that the
>    patch short circuits error conditions\. I've simplified the patch\.
>    I'm pasting the commit message below since it's got all of my
>    thoughts in it&#8203;:
> 
>     RT 76668 \- Check string to compile for chars ~\]\[ and return
>    \\"$string" if not found\.
>     This is a 250% speed improvement on strings which don't require
>    compile and only a
>     ~2% hit if they did need compiling\.
> 
>     Remove \\G since everything is being captured it has no value\. This
>    means we don't
>     have to worry about seting pos $string\_to\_compile = 0 to prevent
>    the previous
>     regex from affecting this one\. There is a negligible speed
>    improvement removing
>     the \\G
> 
> All tests continue to pass\.
> 
> FYI if you want to benchmark on your own\, this is the method I used to
>    show the speedup&#8203;:
> 
> Testing compiled string performance&#8203;:
> perl \-Ilib \-MBenchmark \-MLocale&#8203;::Maketext \-le 'package X;our
>    @&#8203;ISA=qw\(Locale&#8203;::Maketext\);package X&#8203;::en;our @&#8203;ISA=qw\(X\);our %Lexicon
>    = \(\_AUTO => 1\);package main;my $h=X\->get\_handle\(\); my $c = 1;
>    timethis\(100000\, sub\{$h\->maketext\("bar $c \[\_1\]"\, $c\); $c\+\+\}\)'
> 
> Testing strings which don't need compile
> perl \-Ilib \-MBenchmark \-MLocale&#8203;::Maketext \-le 'package X;our
>    @&#8203;ISA=qw\(Locale&#8203;::Maketext\);package X&#8203;::en;our @&#8203;ISA=qw\(X\);our %Lexicon
>    = \(\_AUTO => 1\);package main;my $h=X\->get\_handle\(\); my $c = 1;
>    timethis\(800000\, sub\{$h\->maketext\("bar $c"\); $c\+\+\}\)'
> 

Thank you\. I’ve applied this as 152eae845915a3a33260ad109ad3888c8894d666\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
